cpp/include/cuvs/neighbors/ivf_pq.hpp

```diff
@@ -0,0 +1,413 @@
+             raft::device_matrix_view<const T, IdxT, raft::row_major> dataset)       \
+    ->cuvs::neighbors::ivf_pq::index<IdxT>;                                          \
+                                                                                     \
+  void build(raft::resources const& handle,                                          \
+             const cuvs::neighbors::ivf_pq::index_params& params,                    \
+             raft::device_matrix_view<const T, IdxT, raft::row_major> dataset,       \
+             cuvs::neighbors::ivf_pq::index<IdxT>* idx);                             \
+                                                                                     \
+  auto extend(raft::resources const& handle,                                         \
+              raft::device_matrix_view<const T, IdxT, raft::row_major> new_vectors,  \
+              std::optional<raft::device_vector_view<const IdxT, IdxT>> new_indices, \
+              const cuvs::neighbors::ivf_pq::index<IdxT>& orig_index)                \
+    ->cuvs::neighbors::ivf_pq::index<IdxT>;                                          \
+                                                                                     \
+  void extend(raft::resources const& handle,                                         \
+              raft::device_matrix_view<const T, IdxT, raft::row_major> new_vectors,  \
+              std::optional<raft::device_vector_view<const IdxT, IdxT>> new_indices, \
+              cuvs::neighbors::ivf_pq::index<IdxT>* idx);                            \
+                                                                                     \
+  void search(raft::resources const& handle,                                         \
+              const cuvs::neighbors::ivf_pq::search_params& params,                  \
+              cuvs::neighbors::ivf_pq::index<IdxT>& index,                           \
+              raft::device_matrix_view<const T, IdxT, raft::row_major> queries,      \
+              raft::device_matrix_view<IdxT, IdxT, raft::row_major> neighbors,       \
+              raft::device_matrix_view<float, IdxT, raft::row_major> distances);     \
+                                                                                     \
+  void serialize(raft::resources const& handle,                                      \
+                 std::string& filename,                                              \
+                 const cuvs::neighbors::ivf_pq::index<IdxT>& index);                 \
+                                                                                     \
+  void deserialize(raft::resources const& handle,                                    \
+                   const std::string& filename,                                      \
+                   cuvs::neighbors::ivf_pq::index<IdxT>* index);
+
+CUVS_IVF_PQ(float, int64_t);
```

Comment:
Let's remove all of these macros and instantiate each of these types directly. The macros make it impossible to document these functions. In addition, these functions all need to be documented. Also, we shouldn't have any more templated functions in header files in cuVS- all of the public APIs should already be instantiated for each specific type we support (thus, no need for compile-time instantiations of any templates in the public APIs).